fetch/README.md

```diff
@@ -0,0 +1,134 @@
+- [Spec Proposal Info](#spec-proposal-info)
+- [Wire protocol](#wire-protocol)
+- [Interfaces](#interfaces)
+- [Implementation Details](#implementation-details)
+- [Removing existing relay protocol](#removing-existing-relay-protocol)
+
+## Overview
+
+The fetch protocol is a means for a libp2p node to get data from another based on a given key. It generally fulfills
+the contract:
+
+`Fetch(key) (value, statusCode)`
+
+## Spec Proposal Info
+### Context
+
+Currently there is no standard protocol for doing simple data retrieval. Sometimes a small data retrieval protocol is all
+that's needed to get the job done. This is especially true when it comes to augmenting or layering protocols together.
+
+### Motivation
+
+This specification is helpful because the protocol is already part of a proposal to add a persistence layer on top of
+[PubSub](../pubsub).
+
+There have been numerous discussion revolving around how to allow PubSub to be persistent, some of which go back multiple
+years (more background information is available [here](https://github.com/libp2p/go-libp2p/issues/698)).
+Initial suggestions included modifying or leveraging the existing PubSub spec and implementations.
+However, it was decided that PubSub should remain agnostic to any sort of persistence,
+and that persistence should be layered on top of PubSub as a separate protocol.
+
+There is currently an [IPNS-over-PubSub spec PR](https://github.com/ipfs/specs/pull/218) that describes how the Fetch
+protocol can be used to create a persistence layer on top of PubSub and therefore greatly improve IPNS performance.
+
+### Scope
+
+This is one small protocol that as of right now is only being used to improve the
+[go-libp2p-pubsub-router](https://github.com/libp2p/go-libp2p-pubsub-router). This protocol should not require any one
+who doesn't want to use it to change anything about their current use.
+
+### Potential Effects
+
+1. A persistent pubsub protocol easily built on top of Fetch
+2. A reusable fetch protocol that can be used to augment various other protocols
+(e.g. add the ability to directly ask peers for data to a DHT)
+3. Might be abused/overused given the simple nature of the protocol
+
+### Expected Feature Set and Tentative Technical Directions
+ 
+Should support: `Fetch(key) (value, statusCode)`
+
+However, the level of specificity in the types of the above variables has wiggle room if people are interested.
+The `go-libp2p-pubsub-router` implementation requires:
+ 
+ `key`: At least as generic as a UTF-8 string
+ 
+ `value`: At least as generic as a byte array
+ 
+ `statusCode`: Supports at least `SUCCESS` and `NOT_FOUND`
+
+## Wire protocol
+
+The libp2p protocol ID for this protocol is `/libp2p/fetch/0.0.1`
+
+### Message Format
+The messages in the Fetch protocol use on of the following protobufs (proto3 syntax):
+
+```
+message RequestLatest {
+	string identifier = 1;
+}
+
+message RespondLatest {
+	StatusCode status = 1;
+	enum StatusCode {
```

Comment:
we also need an error code for internal server errors.

Reply:
As discussed related to the go implementation PR on this spec (https://github.com/libp2p/go-libp2p-pubsub-router/pull/33#discussion_r312675975) adding new protobuf enums is really easy and it's not clear anyone would care about actually sending errors over the wire (as opposed to something like just resetting the stream).

If you'd really me to put in an unused `ERR` enum though, that's not particularly bothersome.

Reply:
An `ERROR` is useful for debugging, streams may be closed for other reasons, so please add it.

Reply:
I don't have serious issues with reserving a potentially unused `ERROR` enum in the spec (although we don't have any elements in the spec for when/how to use it, and I don't think we have a use for that yet).

However, if you want to discuss whether the go implementation should utilize the Error, please check out the comment on the PR I linked and voice your opinion there. Both options were explored and implemented, and there are some thoughts there as to why I settled on the no ERROR approach.

Reply:
This is a more general protocol than just your patch! 
There can be situations where you try to fetch data from a datastore, and it fails, and the appropriate response is to return an `ERROR`.

Reply:
There might be. There also might be situations where it's more appropriate to return `[][]byte` then `[]byte`.

Given that this is literally just a lookup you could choose to encode any information you want in the returned `[]byte`, including error information. There's nothing in the protocol, as it currently is, that prevents returning useful errors.

However, if all you want is a reserved enum option, with no detailing what types of errors it's for, we can do that.

Reply:
Let's add an error code but not use it. We can treat all codes >= 2 to mean "error" so we can add more if we need them.

Reply:
Done. There is now `ERROR = 2` in the status code enum. (btw it is also in the go implementation protobuf in case anyone wants to copy the file, but remains unused)